tests/integration/kubernetes/k8s-guest-pull-image-encrypted.bats

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env bats
+# Copyright (c) 2024 IBM Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+load "${BATS_TEST_DIRNAME}/lib.sh"
+load "${BATS_TEST_DIRNAME}/confidential_common.sh"
+load "${BATS_TEST_DIRNAME}/confidential_kbs.sh"
+
+export KBS="${KBS:-false}"
+
+setup() {
+    if is_confidential_hardware; then
+        skip "Due to issues related to pull-image integration skip tests for ${KATA_HYPERVISOR}."
+    fi
+
+    if ! is_confidential_runtime_class; then
+        skip "Test not supported for ${KATA_HYPERVISOR}."
+    fi
+
+    [ "${SNAPSHOTTER:-}" = "nydus" ] || skip "None snapshotter was found but this test requires one"
+
+    setup_common
+    ENCRYPTED_IMAGE="${ENCRYPTED_IMAGE:-ghcr.io/confidential-containers/test-container:multi-arch-encrypted}"
+    DECRYPTION_KEY="${DECRYPTION_KEY:-HUlOu8NWz8si11OZUzUJMnjiq/iZyHBJZMSD3BaqgMc=}"
+    DECRYPTION_KEY_ID="${DECRYPTION_KEY_ID:-ssh-demo}"
+}
+
+function setup_kbs_decryption_key() {
+    decryption_key=$1
+    decryption_key_id=$2
+
+    if [ "${KBS}" = "false" ]; then
+        skip "Test skipped as KBS not setup"
+    fi
+
+    if ! is_confidential_hardware; then
+        kbs_set_allow_all_resources
+    fi
+
+    # Note: the ssh-demo
+    kbs_set_resource_base64 "default" "key" "${decryption_key_id}" "${decryption_key}"
+}
+
+function create_pod_yaml_with_encrypted_image() {
+    image=$1
+
+    local CC_KBS_ADDR
+    export CC_KBS_ADDR=$(kbs_k8s_svc_http_addr)
+    kernel_params_annotation="io.katacontainers.config.hypervisor.kernel_params"
+    kernel_params_value+=" agent.guest_components_procs=confidential-data-hub"
+    kernel_params_value+=" agent.aa_kbc_params=cc_kbc::${CC_KBS_ADDR}"
+
+    # Note: this is not local as we use it in the caller test
+    kata_pod_with_encrypted_image="$(new_pod_config "$image" "kata-${KATA_HYPERVISOR}")"
+    set_node "${kata_pod_with_encrypted_image}" "$node"
+    set_container_command "${kata_pod_with_encrypted_image}" "0" "sleep" "30"
+
+    # Set annotation to pull image in guest
+    set_metadata_annotation "${kata_pod_with_encrypted_image}" \
+        "io.containerd.cri.runtime-handler" \
+        "kata-${KATA_HYPERVISOR}"
+    set_metadata_annotation "${kata_pod_with_encrypted_image}" \
+        "${kernel_params_annotation}" \
+        "${kernel_params_value}"
+
+    add_allow_all_policy_to_yaml "${kata_pod_with_encrypted_image}"
+}
+
+@test "Test that creating a container from an encrypted image, with no decryption key fails" {
+
+    # TODO - there is now delete KBS resource to ensure there is no key, so we need to keep
+    # this test running first to ensure that the KBS doesn't have the resource. An alternative
+    # is to run kbs_set_deny_all_resources, but we don't have a way to reset to the default
```

Comment:
About "we don't have a way to reset to the default policy" made me realize a problem on k8s-confidential-attestation.bats because it calls `kbs_set_deny_all_resources`.  We are lucky (or @stevenhorsman made with intent?) that on `K8S_TEST_SMALL_HOST_UNION` k8s-guest-pull-image-encrypted.bats is listed before k8s-confidential-attestation.bats so the former run first. Certainly we should solve that problem, but not on this PR I suggest.

Reply:
A bit of luck and a bit of judgement. I realised in the discussion with Tobin above that it could be a problem to have multiple of them and the ordering is important, which is a no-no for testing for me, hence the TODO warning, but I agree it isn't very sustainable and we need to ensure that the attestation tests are closer to the bottom.